pkg/validation/validation.go

```diff
@@ -0,0 +1,138 @@
+// Package validation provides UpgradeConfig CR validation tools.
+package validation
+
+import (
+	"fmt"
+	"net/url"
+	"runtime"
+	"time"
+
+	"github.com/blang/semver"
+	"github.com/google/uuid"
+	configv1 "github.com/openshift/api/config/v1"
+	"github.com/openshift/cluster-version-operator/pkg/cincinnati"
+	upgradev1alpha1 "github.com/openshift/managed-upgrade-operator/pkg/apis/upgrade/v1alpha1"
+	"github.com/openshift/managed-upgrade-operator/pkg/osd_cluster_upgrader"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/go-logr/logr"
+)
+
+// NewBuilder returns a validationBuilder object that implements the ValidationBuilder interface.
+func NewBuilder() ValidationBuilder {
+	return &validationBuilder{}
+}
+
+// Validator knows how to validate UpgradeConfig CRs.
+//go:generate mockgen -destination=mockValidation.go -package=validation github.com/openshift/managed-upgrade-operator/pkg/validation Validator
+type Validator interface {
+	IsValidUpgradeConfig(uC *upgradev1alpha1.UpgradeConfig, cV *configv1.ClusterVersion, logger logr.Logger) (bool, error)
+}
+
+type validator struct{}
+
+func (v *validator) IsValidUpgradeConfig(uC *upgradev1alpha1.UpgradeConfig, cV *configv1.ClusterVersion, logger logr.Logger) (bool, error) {
+	// Validate upgradeAt as RFC3339
+	_, err := time.Parse(time.RFC3339, uC.Spec.UpgradeAt)
+	if err != nil {
+		return false, fmt.Errorf("Failed to parse upgradeAt:%s during validation: %v", uC.Spec.UpgradeAt, err)
+	}
+
+	// Validate desired version.
+	dv := uC.Spec.Desired.Version
+	cv, err := osd_cluster_upgrader.GetCurrentVersion(cV)
+	if err != nil {
+		return false, fmt.Errorf("Failed to get current cluster version during validation: %v", err)
+	}
+
+	// Check for valid SemVer and convert to SemVer.
+	desiredVersion, err := semver.Parse(dv)
+	if err != nil {
+		return false, fmt.Errorf("Failed to parse desired version %s as semver: %v", dv, err)
+	}
+	currentVersion, err := semver.Parse(cv)
+	if err != nil {
+		return false, fmt.Errorf("Failed to parse current version %s as semver: %v", cv, err)
+	}
+
+	// Compare versions to ascertain if upgrade should proceed.
+	proceed := compareVersions(desiredVersion, currentVersion, logger)
+	if !proceed {
+		return false, nil
+	}
+	logger.Info(fmt.Sprintf("Desired version %s validated as greater then current version %s", desiredVersion, currentVersion))
+
+	// Validate available version is in Cincinnati.
+	desiredChannel := uC.Spec.Desired.Channel
+	clusterId, err := uuid.Parse(string(cV.Spec.ClusterID))
+	if err != nil {
+		return false, err
+	}
+	upstreamURI, err := url.Parse(string(cV.Spec.Upstream))
+	if err != nil {
+		return false, err
+	}
+
+	updates, err := cincinnati.NewClient(clusterId, nil, nil).GetUpdates(upstreamURI, runtime.GOARCH, desiredChannel, currentVersion)
+	if err != nil {
+		return false, err
```

Comment:
I mentally go back and forth on the Cincinnati stuff and don't have a horse in the race for whether we call it or not, but outside of that:

- If CVO is already reporting the version as valid, do we need to do a Cincinnati check at all? This could help remove Cinny as a blocking dependency in the majority of situations.
- If CVO is not reporting the version as valid, but Cincinnati does, do we still go ahead? (ie. is this a case where urgent CVE must go out, CVO hasn't synced yet, but we know that Cincinnati thinks everything's a-ok)
- If Cincinnati is down, could the error message please indicate that - we are currently relying on the error returned by the cincinnati client to contain useful data, which it may not. so eg
```suggestion
		return false, fmt.Errorf("Failed to pull Cincinnati updates: %v", err)
```

Reply:
i think we leave this in until OCM is in front of the service and is validating against CIS.